README.md
# Incredibots 2 HTML5 Open Source

Early-stage project attempting to port the flash game [IncrediBots 2 CE](http://incredibots.com/if/game.php) to TypeScript and PIXI.js.

## IncrediBots Open Source README from the original developers of the Flash version

*Development environment instructions in what follows do not apply to the HTML5 version.*

## PLAYING THE GAME

Thank you for downloading IncrediBots!  Get ready to build, control, and share your
very own robot creations in a physics-based playground!

To run IncrediBots or IncrediBots2, simply open the file incredibots.html or
incredibots2.html in a web browser with flash player installed.

Enjoy!

## BUILDING FROM SOURCE

The source code for the IB1 and IB2 game clients are located in the src/Incredibots
and src/Incredibots2 subdirectories.  Open up the actionscript IDE of your choice
(I used Flex builder 3), and create a new project using the appropriate directory.
The main runnable application file should be set to Main.as, and I used the param
-default-frame-rate 30 as a compiler argument.  Also make sure to link to the
flGUI.swc library in the libs folder.  The source code contains a few [Embed]
statements, that I believe may only work with Flex builder, but other IDEs likely
have similar methods of linking to assets.

Once everything is set up and the project compiles properly, the IDE should create
a bin-debug folder (or something similar), containing the SWF and an HTML file that
embeds the SWF.  Now you'll be able to make changes to the code and see them
reflected in the built SWF!

## CLIENT SIDE CODE

I'll give a basic overview of the key classes in the client side code, in hopes of
making it clear how exactly everything works.  The main application class is Main.as,
which takes care of adding event listeners to the stage, and determining which
'Controller' class is active.  You can think of a controller as loosely representing
one type of 'screen.'  There's one for the main menu, and then one for the main game
screen, with subclasses for all the different types of main game screens (i.e 1
subclass per tutorial, 1 for sandbox mode, 1 for create challenge mode, etc.).  The
(unfortunately somewhat monolithic) class that handles most of the main robot editing
functionality is ControllerGame, which has methods for responding to all the buttons
and menu options that are available while editing a robot, as well as methods for
handling creating, deleting, selecting, moving, and editing robot parts.  The allParts
Array in ControllerGame represents the robot currently being worked on, as an Array
of Parts.  It also handles starting and stopping the Box2D simulation

The Parts package contains class definitions for each type of 'Part available when
building a robot.  i.e. Circle, Rectangle, Fixed Joint, Rotating Joint, Thrusters, etc.
Each instance of these classes contains all the data about that particular part, for
example, a RevoluteJoint has member variables for joint strength, speed, whether or
not the motor is enabled, control keys, which 2 shape parts it's attached to, etc.
In the code, a 'Robot' or a 'Challenge' can be thought of as just an array of parts
with a bit of metadata attached (i.e. sandbox settings, challenge restrictions).

The Actions package simply contains a list of actions that can be performed by the
user when in edit mode.  These actions are used solely by the 'Undo' and 'Redo'
features, and each action knows what needs to be done to undo or redo it.

The Input class handles all keyboard and mouse interactions with non-GUI elements,
and dispatches calls to the appropriate methods (usually in ControllerGame) to
respond to a given mouse click or key press.

The Draw class deals with the drawing of robots to the screen.  There are 2 cases
it needs to deal with: edit mode, and simulation mode.  When in edit mode, it draws
all the shapes to the screen based on the current Array of Parts.  When in sim mode,
it draws the shapes based on the positioning of the Box2D representation of those
parts, but it needs to make them look the same in both cases.  In theory, this class
is pretty modular, and should be able to be swapped out with an alternate class used
for drawing to the screen fairly easily, in the event that you'd like to write your
own methods for dealing with robot graphics.

The Gui package has a bunch of classes that build the various GUI panels and windows
throughout the game.  The most important of these is probably the SaveLoadWindow
class, which handles displaying a list of robots available for load from the server.
It also handles high scores, saving and loading robots, replays, and challenges, and
searching for robots and such.  It's currently disabled, since the servers are going
down, but if another server ever goes up, this class will have to be modified.  Other
classes in this package include GuiWindow, which is the base class for all the little
popup dialogs throughout the game, and components such as GuiButton, GuiTextArea, etc.

Finally, the Database class handles all interactions with the server.  It has operations
for saving and loading robots and replays, as well as for getting lists of each of
these, and for logging in, creating a new user, uploading/downloading scores, and a few
more.  Each operation has a corresponding callback function, which processes the result
returned by the server and returns the data back to the appropriate class for display.
Of note are the functions PutXIntoByteArray and ExtractXFromByteArray, where X is one
of 'Robot', 'Replay', or 'Challenge'.  These functions are used to read and write in-game
data to or from a flash ByteArray.  The robot, replay, or challenge is then sent to the
server as the POST data part of a URLRequest.  These functions are also used by the
'Import' and 'Export' features, as an easy way of converting between in-game objects and
text.  If you want to add features that need to be saved along with a robot, replay, or
challenge, you'll most likely need to put them in to one of these functions explicitly.

## SERVER SIDE CODE

The server code is everything contained within the htdocs subfolder.  It's written
entirely in PHP.  I'll talk about the database server code first.

The database code is in the ./src/htdocs/incredibots[1 or 2]/database/ directory.
IncrediBots was build using Amazon S3 and SimpleDB as backends.  You can read more
about these at [http://aws.amazon.com](http://aws.amazon.com) .  Most of the DB server code is written
specifically with this in mind, and unfortunately I never got around to refactoring
it such that the database layer is abstracted within a common package.  Thus it may
be a lot of work to switch to something like MySQL, although it would work just as
well as SimpleDB and S3 if properly maintained.

Note that IncrediBots also makes use of MySQL through the PHPBB forums database.
Because we had a single login for both the game and the forums, all user accounts
are managed through the forums' phpbb_users table.  This is why the PHPBB forums
code is included in the package, I had to make a few modifications to the forums
code to get the single login working, as well as a few other features, such as the
ability to embed robots within forum posts.  That said, basic login and creating new
accounts should work just fine with any version of phpbb running, and it shouldn't be
too difficult to change the login functionality to work with whatever database/system
you want.

The current database code should work fine out of the box with any Amazon AWS account.
The only things that should need modifying are the credentials in the file
common_variables.php.  Apart from that, each ggscores_*.php file corresponds to one
of the 20 or so DB operations the client can call.  All calls are initially directed
to ggscores.php, which includes the proper file corresponding to the 'op' parameter.
Each operation then talks to the Amazon database (or phpbb_users table in MySQL) and
echos out a response for the client.  The binary data for all robots, replays, and
challenges is sent to the server via HTTP POST data, and are saved in S3 as BLOBs,
with a small amount of metadata for each stored in SimpleDB.

NOTE: One thing it may be a good idea to add is an MD5 checksum validation when
sending the binary data from the client to the server, to guard against packet
corruption.  I know this has happened in the past a few times, and when it does, the
affected robot/replay/challenge becomes corrupt, and is lost forever...  (Dun dun dun!)

The rest of the server code contains mostly components of the IncrediBots website.
The index.php page contains code to embed the SWF and display it properly.  In order
to host a SWF, you would just upload the SWF file that results from building the
client, named incredibots.swf (or incredibots2.swf) to the same directory as index.php.
The users.php page is also here, which displays all robots, replays, and challenges
uploaded by a given user in a (somewhat) nicely formatted webpage.

## ABOUT

IncrediBots was originally designed by Ryan Clark, programmed by Oliver Trujillo,
with art from Matt Parry.  A big amount of credit should also go to Erin Catto and
the Box2D physics engine, as well as the Flash port of it upon which our game is
built.
[http://www.box2d.org/](http://www.box2d.org/)
[http://box2dflash.sourceforge.net/](http://box2dflash.sourceforge.net/)

> IncrediBots is probably the game I've had the most fun working on.  It's been a
> blast checking out all the crazy creations made by the IncrediBots community, and I
> wish you all the best in keeping IncrediBots going.  Cheers and happy bot building!
> - Oliver


CHANGELOG.md
# IncrediBots2 Changelog

## Version 2.23 CE HTML5

- WIP port of 2.23CE to HTML5

## Version 2.23 CE

- Removed Cheat Engine restriction added in 1.08/2.11 and "forked" development to a new edition called "CE" (for "Cheat Engine"). Note that while bots, challenges, and replays can be loaded in between CE and non-CE Versions, any CE-specific modifications (i.e. out-of-range values) will not work in non-CE Versions. This includes replays.
Uploaded Jan 24, 2010

## Version 2.23

- Fixed bugs in source code that prevented successful importing of robots and challenges in both Incredibots 1 and 2 (pre-compiled SWFs are fine).
- Also fixed both IncrediBots 1 and 2 mouse cursor bugs (Incredibots 1 with its reappearing OS cursor, Incredibots 2 with its disappearing/flickering OS cursor). Bug fixes in source code only (pre-compiled SWFs are untouched).
Uploaded Nov 24, 2010

## Version 2.22

- shut IncrediBots 1 and 2 websites down and uploaded open-sourced Versions to SourceForge (http://sourceforge.net/projects/incredibots/)
Uploaded Nov 19, 2010

## Version 2.14

- disabled saving, and added importing/exporting to text
Uploaded Oct. 20, 2010

## Version 2.13

- fixed bug with challenge featuring
Uploaded Aug. 12, 2009

## Version 2.12

- changes to featuring system
Uploaded July 6, 2009

## Version 2.11

- various minor updates and bug fixes
Uploaded June 25, 2009

## Version 2.10

- IncrediBots 2 released publicly!
- added ability to export your robots from IB1 to IB2. Note that the physics will not be exactly the same, so rube goldberg machines and such may need some tweaking.
- fixed a few bugs involving using cannons in challenges and in replays
- tweaked sounds a bit
- added 4 challenges to the tutorial menu
Uploaded June 19, 2009

## Version 2.08

- added sound!
- added cannons!
- changed default color and colors in the drop down box
- selected shapes now turn transparent
- fixed bug with displayable text not working in replays
- fixed duplicate featured stuff bug
- UI improvements to login/logout
Uploaded May 28, 2009

## Version 2.07

- added default color feature to color change GUI
- various tutorial and UI improvements
Uploaded May 14, 2009

## Version 2.06

- added 2 new options to restrictions GUI
- added option to turn off collisions for sliding joints
- fixed bug with text disappearing in challenges
- added 2 new tutorials explaining some new features in IB2
Uploaded May 1, 2009

## Version 2.05

- fixed another bug with certain loss conditions
- fixed bug with slow rotating objects being jerky in replays
- robots can now be saved for specific challenges, and the challenge will load when the robot loads
- added undraggable feature (thanks to those that suggested this!)
- added multi-fixate and multi-collide features
- fixed mirroring sliding joint bug
Uploaded April 21, 2009

## Version 2.04

- fixed bugs having to do with challenge conditions
- fixed many other bugs that were reported in the beta testing thread
- added a few buttons/menu options here and there
Uploaded April 16, 2009

## Version 2.01

- fixed bug that made challenged uneditable
- fixed bug that crashed browser when using certain win/loss conditions
- fixed bug that made multi-select box stay visible after selecting things
- fixed several bugs that caused weird results after loading and stopping replays of challenges
- added feature that saves latest camera position and zoom level of robots and challenges
Uploaded Mar. 19, 2009

## Version 2.00

- Incredibots 2 released for private beta testing!
- added challenges, new sandbox features, mouse dragging, and more
- see post in Feature Requests forum for more details.
Uploaded Mar. 17, 2009


src/Main.ts
// permanent mochi ID for version 0.02: 1913f89f65e17063

import { Application, Container, Sprite } from "pixi.js";
import { Controller } from "./Game/Controller";
import { ControllerGameGlobals } from "./Game/Globals/ControllerGameGlobals";
import { ControllerMainMenu } from "./Game/ControllerMainMenu";
import { ControllerSandbox } from "./Game/ControllerSandbox";
import { SandboxSettings } from "./Game/SandboxSettings";
import { FpsCounter } from "./General/FpsCounter";
import { LSOManager } from "./General/LSOManager";
import { Input } from "./General/Input";

export class Main {

	//======================
	// Member data
	//======================
	public static m_fpsCounter:FpsCounter = new FpsCounter();
	public static m_curController:Controller;
	public static changeControllers:boolean = false;
	public static nextControllerType:number = -1;
	public static firstFrame:boolean = true;
	public static frameCounter:number = 0;
	public static inIFrame:boolean = false;

	public static renderer: Application;
	public static theRoot:Container;

	public static mouseCursor:BitmapAsset;
	public static mouseHourglass:BitmapAsset;

	public static loadRobotMode:boolean = false;
	public static loadReplayMode:boolean = false;
	public static loadChallengeMode:boolean = false;
	public static premiumMode:boolean = false;
	public static enableSound:boolean = true;

	public static const VERSION_STRING:string = "2.23 CE";
	public static const DEBUG_VERSION:boolean = true;
	public static const ENABLE_SITE_LOCK:number = 643 - (Main.DEBUG_VERSION ? 2 : 1);
	public static _mochiads_game_id:string = "50feb50977299858";

	public static lastAdTime:number = 0;

	public static const GLOBAL_FONT:string = "Arial";

	public static BrowserRedirect(url:string|null = null, newWindow:boolean = false, parent:boolean = false):void {
		const target = url || "http://www.incredibots.com/"
		let mode = '_self'

		if (newWindow) mode = '_blank'
		if (parent) mode = '_top'

		window.open(target, mode)
	}

	constructor(renderer: Application) {
		Main.renderer = renderer
		Main.theRoot = Main.renderer.stage
		Main.theRoot.interactive = true

		var urlString:string = window.location.href;
		var urlStart:number = urlString.indexOf("://") + 3;
		var urlEnd:number = urlString.indexOf("/", urlStart);
		var domain:string = urlString.substring(urlStart, urlEnd);
		var lastDot:number = domain.lastIndexOf(".") - 1;
		var domEnd:number = domain.lastIndexOf(".", lastDot) + 1;
		domain = domain.substring(domEnd, domain.length);

		if (domain == "incredibotsgold.com") Main.premiumMode = true;

		renderer.ticker.maxFPS = 30 // Cap FPS to 30 as the FPS is tied directly to the game loop and the flash version was designed this way. :(
		renderer.ticker.add((delta) => this.update())

		addEventListener(Event.ADDED_TO_STAGE, this.Init, false, 0, true);

		LSOManager.Init();
		this.Init();

		if (Main.DEBUG_VERSION) {
			Main.m_fpsCounter.x = 7;
			Main.m_fpsCounter.y = 525;
			Main.theRoot.addChild(Main.m_fpsCounter);
		}

		// contextMenu = new ContextMenu();
		// contextMenu.hideBuiltInItems();
	}

	private Init():void {
		Input.Init();
	}

	private rehideMouse(e:Event):void {
		Mouse.hide();
	}

	public update():void {
		Main.frameCounter++;
		if (Main.firstFrame) {
			var urlParams:URLSearchParams = new URLSearchParams(window.location.search);
			if (urlParams.get("replayID")) {
				Main.loadReplayMode = true;
				var replayID:string = urlParams.get("replayID");
			}
			if (urlParams.get("robotID")) {
				Main.loadRobotMode = true;
				var robotID:string = urlParams.get("robotID");
			}
			if (urlParams.get("challengeID")) {
				Main.loadChallengeMode = true;
				var challengeID:string = urlParams.get("challengeID");
			}
			if (urlParams.get("iframe")) {
				Main.inIFrame = true;
			}

			if (Main.loadReplayMode && !Main.premiumMode) {
				ControllerSandbox.settings = new SandboxSettings(15.0, 0, 0, 0, 0);
				Main.m_curController = new ControllerSandbox();
				(Main.m_curController as ControllerGame).LoadReplayNow(replayID);
				ControllerGameGlobals.potentialReplayID = replayID;
				ControllerGameGlobals.potentialReplayPublic = true;
				ControllerGameGlobals.replayDirectlyLinked = true;
				Main.theRoot.addChild(Main.m_curController);
				Main.nextControllerType = 0;
			} else if (Main.loadRobotMode && !Main.premiumMode) {
				ControllerSandbox.settings = new SandboxSettings(15.0, 0, 0, 0, 0);
				Main.m_curController = new ControllerSandbox();
				(Main.m_curController as ControllerGame).LoadRobotNow(robotID);
				ControllerGameGlobals.potentialRobotID = robotID;
				ControllerGameGlobals.potentialRobotPublic = true;
				Main.theRoot.addChild(Main.m_curController);
				Main.nextControllerType = 0;
			} else if (Main.loadChallengeMode && !Main.premiumMode) {
				ControllerSandbox.settings = new SandboxSettings(15.0, 0, 0, 0, 0);
				Main.m_curController = new ControllerSandbox();
				(Main.m_curController as ControllerGame).LoadChallengeNow(challengeID);
				ControllerGameGlobals.potentialChallengeID = challengeID;
				ControllerGameGlobals.potentialChallengePublic = true;
				Main.theRoot.addChild(Main.m_curController);
				Main.nextControllerType = 0;
			} else {
				Main.m_curController = new ControllerMainMenu();
				Main.theRoot.addChild(Main.m_curController);
			}

			// Main.mouseCursor = new Resource.cMouseCursor();
			// Main.mouseCursor.smoothing = true;
			// Main.mouseCursor.visible = (Main.m_curController is ControllerMainMenu);
			// stage.addChild(Main.mouseCursor);
			// Main.mouseHourglass = new Resource.cMouseHourglass();
			// Main.mouseHourglass.visible = !(Main.m_curController is ControllerMainMenu);
			// Main.mouseHourglass.smoothing = true;
			// stage.addChild(Main.mouseHourglass);

			Main.firstFrame = false;
		}

		//trace(getTimer());

		// Reset
		if (Main.changeControllers) {
			Main.theRoot.removeChild(Main.m_curController);
			if (Main.nextControllerType == -1) {
				Main.m_curController = new ControllerMainMenu(true);
			} else if (Main.nextControllerType == 10) {
				Main.m_curController = new ControllerTank();
			} else if (Main.nextControllerType == 11) {
				Main.m_curController = new ControllerShapes();
			} else if (Main.nextControllerType == 12) {
				Main.m_curController = new ControllerCar();
			} else if (Main.nextControllerType == 13) {
				Main.m_curController = new ControllerJumpbot();
			} else if (Main.nextControllerType == 14) {
				Main.m_curController = new ControllerDumpbot();
			} else if (Main.nextControllerType == 15) {
				Main.m_curController = new ControllerCatapult();
			} else if (Main.nextControllerType == 16) {
				Main.m_curController = new ControllerHomeMovies();
			} else if (Main.nextControllerType == 17) {
				Main.m_curController = new ControllerRubeGoldberg();
			} else if (Main.nextControllerType == 18) {
				Main.m_curController = new ControllerNewFeatures();
			} else if (Main.nextControllerType == 19) {
				Main.m_curController = new ControllerChallengeEditor();
			} else if (Main.nextControllerType == 1) {
				Main.m_curController = new ControllerChallenge();
			} else if (Main.nextControllerType == 2) {
				Main.m_curController = new ControllerMonkeyBars();
			} else if (Main.nextControllerType == 3) {
				Main.m_curController = new ControllerClimb();
			} else if (Main.nextControllerType == 4) {
				Main.m_curController = new ControllerRace();
			} else if (Main.nextControllerType == 5) {
				Main.m_curController = new ControllerSpaceship();
			} else {
				Main.m_curController = new ControllerSandbox();
			}
			Main.theRoot.addChild(Main.m_curController);

			Main.changeControllers = false;
		}

		// update current test
		Main.m_curController.Update();

		// Update input (last)
		Input.update();

		//trace(getTimer() + "\n");

		// update counter and limit framerate
		if (Main.DEBUG_VERSION) Main.m_fpsCounter.update();
		//FRateLimiter.limitFrame(30);
	}

	public static ShowMouse():void {
		// Main.mouseHourglass.visible = false;
		// Main.mouseCursor.visible = true;
	}

	public static ShowHourglass():void {
		// Main.mouseCursor.visible = false;
		// Main.mouseHourglass.visible = true;
	}
}
